File: 2020/Регион/Тур4/B/b.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using ld = long double;
using ull = unsigned long long;
#define all(x) x.begin(), x.end()

template <typename T1, typename T2> inline void chkmin(T1 &x, const T2 & y) {if (x > y) x = y;}
template <typename T1, typename T2> inline void chkmax(T1 &x, const T2 & y) {if (x < y) x = y;}

const int MAXN = 110, MAXLEN = 5;
#define int ll
map<string, ll> dp[MAXN];
string s;
int n;

void read() {
	cin >> s >> n;
}

ll ans;

string get(string s, int l, int r) {
	string ans = "";
	for (int i = l; i <= r; i++)
		ans += s[i];
	return ans;
}

bool check(string s, string t) {
	for (int i = 0; i + s.size() - 1 < t.size(); i++) {
		if (get(t, i, i + s.size() - 1) == s)
			return false;
	}
	return true;
}

bool check(string t) {
	string t1 = t;
	reverse(all(t1));
	return check(s, t) && check(s, t1);
}

int fn = 0;

bool check_ans(string t) {
	string fans = t;
	if (fn & 1) t.pop_back();
	reverse(all(t));
	fans += t;
	return check(s, fans);
}

void run() {
	for (int i = 0; i <= 1; i++) {
		string x = "";
		x += 'A' + i;
		if (x != s)
			dp[1][x] = 1;
	}
	fn = n;
	n = (n + 1) / 2;
	for (int i = 1; i < n; i++) {
		for (auto [x, y] : dp[i]) {
			for (int j = 0; j < 2; j++) {
				string now = x;
				now += 'A' + j;
				if (now.size() > MAXLEN)
					now.erase(now.begin());
				if (check(now))
					dp[i + 1][now] += y;
			}
		}
	}

	ans = 0;
	for (auto [s, cnt] : dp[n]) {
		if (check_ans(s))
			ans += cnt;
	}
}

void write() {
	cout << ans << endl;
}

signed main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);
	cout.tie(0);
	read();
	run();
	write();
	return 0;
}

File: 2020/Регион/Тур4/B/b.py
def get_substr(s, l, r):
	ans = ""
	for i in range(l, r + 1):
		ans += s[i]
	return ans

def get_border(s, t):
	n = len(s)
	if (len(t) < len(s)):
		n = len(t)
	ans = 0
	for i in range(n):
		if (get_substr(s, len(s) - 1 - i, len(s) - 1) == get_substr(t, 0, i)):
			ans = i + 1
	if (ans == len(s)):
		return -1
	return ans;

c = ["A", "B"]

def build(s):
	ans = []
	x0 = 0
	if (s[0] == "A"):
		x0 = 1
	x1 = x0 ^ 1
	ans.append([x0, x1])
	now = ""
	for i in range(len(s) - 1):
		now += s[i]
		add = []
		for j in range(2):
			add.append(get_border(now + c[j], s))
		ans.append(add)
	return ans

s = input()
n = int(input())

dp = []
for i in range(n + 1):
	x = []
	for j in range(len(s)):
		x.append(0)
	dp.append(x)

dp[0][0] = 1
#print(dp)

nxt = build(s)
print("Nxt =", nxt)
x = (n + 1) // 2
#print("x =", x)
for i in range(x):
	for j in range(0, len(s)):
		for k in range(2):
			if (dp[i][j] == 0 or nxt[j][k] == -1):
				continue
			dp[i + 1][nxt[j][k]] += dp[i][j]

print("dp =", dp)
ans = 0

def check(s, t):
	i = 0
	while i + len(t) - 1 < len(s):
		if (get_substr(s, i, i + len(t) - 1) == t):
			return False
		i += 1
	return True



def Check(Len):
	x = s[:Len + 1]
	y = x[::-1]
	if n % 2 == 1:
		y = y[1:]
	fans = x + y
	print(fans, s)
	return check(fans, s)

for i in range(0, len(s)):
	if (Check(i)):
		ans += dp[x][i]
print(ans)










